src/Compilers/Core/Portable/CommandLine/ErrorLogger.cs

```diff
@@ -88,48 +86,62 @@ internal void LogDiagnostic(Diagnostic diagnostic, CultureInfo culture)
 
         private void WriteLocations(Location location, IReadOnlyList<Location> additionalLocations)
         {
-            _writer.WriteArrayStart("locations");
+            if (location.SourceTree != null)
+            {
+                _writer.WriteArrayStart("locations");
+                _writer.WriteObjectStart(); // location
+                _writer.WriteKey("analysisTarget");
```

Comment:
It's good that MSBUILD's behavior can be accommodated by your position, but there may be scenarios where that isn't the case. It isn't ideal to say that SARIF only works in this case if you make sure to produce log files on a per-component basis. I agree the project file idea wasn't a good proposal.

analysisTarget was intended to represent a discrete scan target. in some cases (such as a C++ object file) a scan target is an aggregate of files which might produce a result. viewed this way, an individual C# file is not an analysis target. the compilation is the scan. the source file is the result file, one of the individual things that were inspected as part of the higher level 'invoke roslyn for this component'. so I stand by my position.

perhaps you will be persuaded by considering the binskim/fxcop vs. roslyn case. when you drive a roslyn analyzer through binskim (or you run equivalent rules in fxcop), analysis target will be populated by the assembly path and file locations will appear as result files. when you drive the same rules against the same target as part of a Roslyn compilation, your current implementation will move all the results to analysisTarget location. again, I think the right thing here is to populate resultFile instead (and not to populate analysisTarget at all if we can't think of what should be there).